framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,605 @@
+        StringBuilder sb = new StringBuilder();
+        for (Annotation tla : annos) {
+            if (!tla.def.name.contains("+")) {
+                sb.append(formatAnnotation(tla));
+                sb.append(" ");
+            }
+        }
+        return sb.toString();
+    }
+
+    /**
+     * Formats the annotations on the component type of an array, if there are any.
+     *
+     * @param e the array type
+     * @return the array type formatted to be written to Java source code
+     */
+    private String formatArrayComponentTypeAnnotation(ATypeElement e) {
+        StringBuilder result = new StringBuilder();
+        for (Map.Entry<InnerTypeLocation, ATypeElement> ite : e.innerTypes.entrySet()) {
+            InnerTypeLocation loc = ite.getKey();
+            AElement it = ite.getValue();
+            if (loc.location.contains(TypePathEntry.ARRAY)) {
+                result.append(formatAnnotations(it.tlAnnotationsHere));
+            }
+        }
+        return result.toString();
+    }
+
+    /**
+     * Formats an AField so that it can be printed in a stub. An AField represents a variable
+     * declaration. In practice, {@code aField} should represent either a field declaration or a
+     * formal parameter of a method, because stub files should not contain local variable
+     * declarations. It is the responsibility of each caller of this method to ensure that is true.
+     *
+     * @param aField the field to format
+     * @param fieldName the name to use for the declaration in the stub file. This doesn't matter
+     *     for parameters, but must be correct for fields.
+     * @param className the name of the enclosing class. This is only used for printing the type of
+     *     an explicit receiver parameter (i.e., a parameter named "this").
+     * @return a String suitable to print in a stub file
+     */
+    private String formatAField(AField aField, String fieldName, String className) {
+        StringBuilder result = new StringBuilder();
+        String basetype;
+        if ("this".equals(fieldName)) {
+            basetype = className;
+        } else if (basetypes.containsKey(aField.type.description.toString())) {
+            basetype = basetypes.get(aField.type.description.toString()).toString();
+        } else {
+            throw new BugInCF(
+                    "SceneToStubWriter: could not find the base type for this variable declaration: "
+                            + fieldName);
+        }
+
+        if (basetype.contains("[")) {
+            String component = basetype.substring(0, basetype.lastIndexOf('['));
+            result.append(formatArrayComponentTypeAnnotation(aField.type));
+            result.append(component);
+            result.append(" ");
+            basetype = "[]";
+        }
+        result.append(formatAnnotations(aField.type.tlAnnotationsHere));
+        result.append(basetype);
+        result.append(' ');
+        result.append(fieldName);
```

Comment:
It's weird that this does not include a semicolon.  I would have this method create a syntactic entity rather than forcing clients to add the semicolon.

Reply:
This method prints both field definitions and formal parameters, so printing a semicolon is inappropriate. I added another paragraph to the documentation to make it very clear.